pkg/config/expression.go

```diff
@@ -64,47 +67,216 @@ func SimpleVarToReference(s string) (Reference, error) {
+	case "var":
+		n, err := getAttrName(1)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected second component of global var reference to be a variable name, got %w", err)
+		}
+		return Reference{GlobalVar: true, Name: n}, nil
+	case "module":
+		m, err := getAttrName(1)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected second component of module var reference to be a module name, got %w", err)
+		}
+		n, err := getAttrName(2)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected third component of module var reference to be a variable name, got %w", err)
+		}
+		return Reference{Module: m, Name: n}, nil
+	default:
+		return Reference{}, fmt.Errorf("unexpected first component of reference: %#v", root)
+	}
+}
+
+// IsRawHclLiteral checks if passed value of type cty.String
+// and its content starts with "((" and ends with "))".
+// Returns trimmed string and result of test.
+func IsRawHclLiteral(v cty.Value) (string, bool) {
+	if v.Type() != cty.String {
+		return "", false
+	}
+	s := v.AsString()
+	if len(s) < 4 || s[:2] != "((" || s[len(s)-2:] != "))" {
+		return "", false
+	}
+	return s[2 : len(s)-2], true
+}
+
+// HclExpression is a representation of HCL literals in Blueprint
+type HclExpression struct {
+	e  hclsyntax.Expression
+	s  string
+	rs []Reference
+}
+
+// ParseExpression returns HclExpression
+func ParseExpression(s string) (HclExpression, error) {
+	e, diag := hclsyntax.ParseExpression([]byte(s), "", hcl.Pos{})
+	if diag.HasErrors() {
+		return HclExpression{}, diag
+	}
+	ts := e.Variables()
+	rs := make([]Reference, len(ts))
+	for i, t := range ts {
+		var err error
+		if rs[i], err = TraversalToReference(t); err != nil {
+			return HclExpression{}, err
+		}
+	}
+	return HclExpression{e: e, s: s, rs: rs}, nil
+}
+
+// MustParseExpression is "errorless" version of ParseExpression
+// NOTE: only use it if passed expression is guaranteed to be correct
+func MustParseExpression(s string) HclExpression {
+	if exp, err := ParseExpression(s); err != nil {
+		panic(fmt.Errorf("error while parsing %#v: %w", s, err))
+	} else {
+		return exp
+	}
+}
+
+// Tokenize returns Tokens to be used for marshalling HCL
+func (e HclExpression) Tokenize() hclwrite.Tokens {
+	return hclwrite.TokensForIdentifier(e.s)
+}
+
+// References return Reference for all variables used in the expression
+func (e HclExpression) References() []Reference {
+	c := make([]Reference, len(e.rs))
+	for i, r := range e.rs {
+		c[i] = r
+	}
+	return c
 }
 
-// DoNotAllowVariablesTranslator does not do any translation, it raises an error if any variables are met
-type DoNotAllowVariablesTranslator struct {
-	VariableTranslator
+// AsRawValue returns a cty.Value, that is rendered as
+// HCL literal in Blueprint syntax. Returned value isn't functional,
+// as it doesn't reference HclExpression.
+// This method should only be used for marshaling Blueprint YAML.
+func (e HclExpression) AsRawValue() cty.Value {
```

Comment:
This comment is very good but it's also worth getting name of this right. What do we mean by "Raw"? Is "Value" the right word?

- "For the blueprint"
- "has (( and ))"?

I have experienced a lot of confusion over time trying to figure out what we mean in various contexts.

Would it be helpful to name it "AsBlueprintLiteral"?

https://github.com/GoogleCloudPlatform/hpc-toolkit/tree/main/examples#literal-variables

I have never liked calling them "literal variables" and prefer "literals".

Reply:
Renamed to `makeYamlLiteralValue`.